Content/RootScript/BaseClasses/GUI/GUITextEntry.as

```diff
@@ -0,0 +1,183 @@
+
+// GUITextEntry --------
+// Text Input beast.
+// One line only.
+// Text boxes maybe another day, game doesn't need it.
+
+namespace GUI { namespace TextEntry {
+
+} }
+
+shared Font@ tempFont=Font::large;
+
+shared class GUICharacter : GUIClickable {
+	// A single character.
+	// Yes, I know.
```

Comment:
No, you don't. This is not permissible.

Reply:
This saves doing vector math over the string to determine clickables, what sections of the string are selected, and all that stuff. This is used exclusively for TextEntry. I agree it is horrendously inefficient, however it is necessary to have a functioning TextEntry gui element. I agree, this is something that can be significant improved in terms of optimization in the future. Worthwhile note; Text entry elements are only ever drawn while they are visible, and thus this character repeat-draws are used in niche cases. A way to cheaply get past this issue can be to compile the characters into a single string and draw that in one draw, so that individual character drawing *only* happens while the textentry element has focus.

Reply:
On further thought, it is actually appropriate to do this for the text entry. Selection, click tracking, Caret, and all that stuff, requires management of each individual character and their positions. the docking system providers a container that already does all of that labor. It is more effective to use an individual character element for each character in this specific scenario, and thus, an individual draw for each character isn't too bad of a trade-off for its ease of use. Again, efficiency can be boosted by compiling the label into a single draw when it is not focused.

Reply:
No, this is not necessary, refer to the current text input implementation.

Reply:
I have found a case that makes GUICharacter not work; Clicking at the end of the bar and not on a character, then selection dragging to the start. Either way, all behavior relating to the text entry bar is contained entirely within the textentry component. At the moment, this method "works", but is also missing many functions like arrow keys, and yes a better draw method. This may remain an open issue until it is finished because I would like to move onto other tasks. Text entry is not used in many places, so a poor functioning text entry is not a major priority. Besides, this implementation does not cause a crash when backspace is pressed, so it is automatically better than the previous textentry implementation. I have put a copy of the old GUI code in the OldSrc folder so it can be retained until this element is finished.

Reply:
Textentry is done correctly now.